### PR TITLE
Classification Evaluation

### PR DESCRIPTION
**Pull request by Negiiiin**
# PR Type
[Feature]

# Short Description
Added zero-shot and linear-probing classification evaluation (accuracy, f1-score and AUC)

- - -
This change is [Reviewable](https://reviewable.io/reviews/VectorInstitute/mmlearn/17)

### CODE REVIEW

File: mmlearn/tasks/classification.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+class ClassificationTaskSpec:
+    """Specification for a classification task."""
+
+    metric_name: str
+    query_modality: str
+    top_k: List[int]
+
+
+@store(group="eval_task", provider="mmlearn")
+class Classification(EvaluationHooks):
```
Comment (fcogidi) [inline mmlearn/tasks/classification.py:34]:
```suggestion
class ZeroShotClassification(EvaluationHooks):
```

File: mmlearn/tasks/classification.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        A list of classification task specifications.
+    """
+
+    def __init__(
+        self,
+        task_specs: List[ClassificationTaskSpec],
+        tokenizer: Optional[
+            Callable[[str], Union[torch.Tensor, Dict[str, torch.Tensor]]]
+        ] = None,
+        mode: Literal["zeroshot", "linear_probing"] = "zeroshot",
```
Comment (fcogidi) [inline mmlearn/tasks/classification.py:52]: I think linear probing should be a separate task. It is not really a zero-shot task (the linear classifier needs to be trained), so it shouldn't inherit from `EvaluationHooks`.

File: mmlearn/tasks/classification.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+
+@store(group="eval_task", provider="mmlearn")
+class Classification(EvaluationHooks):
+    """
+    Zero-shot classification evaluation task.
+
+    This task evaluates the zero-shot classification performance.
+
+    Parameters
```
Comment (fcogidi) [inline mmlearn/tasks/classification.py:40]: Reminder to update this with all the parameters in the `__init__` method.

File: mmlearn/tasks/classification.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+                    collate_fn=collate_example_list,
+                )
+                batch = next(iter(dataloader))
+                batch = {
+                    key: value.to(pl_module.device) if torch.is_tensor(value) else value
+                    for key, value in batch.items()
+                }
+                all_dataset_info[name].set_label_embedding(
+                    pl_module(batch)[Modalities.get_modality(Modalities.TEXT).embedding]
+                )
```
Comment (fcogidi) [inline mmlearn/tasks/classification.py:129]: I don't think all of this is necessary. You could take the class labels as input, tokenize them and get their embedding without having to define a dataset class or dataloader.

File: mmlearn/tasks/classification.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+                                                "num_classes": num_classes,
+                                            }
+                                            if "top_k"
+                                            in inspect.signature(metric).parameters
+                                            else {
+                                                "task": "multiclass",
+                                                "num_classes": num_classes,
+                                            }
+                                        )
+                                    ).to("cuda")
```
Comment (fcogidi) [inline mmlearn/tasks/classification.py:166]:
```suggestion
                                    ).to(self.device)
```

File: mmlearn/tasks/classification.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+        self.all_dataset_info = all_dataset_info
+
+        for dataset_name in list(all_dataset_info.keys()):
+            for spec in self.task_specs:
+                assert Modalities.has_modality(spec.query_modality)
+                query_modality = Modalities.get_modality(spec.query_modality)
+                mode = self.mode
+                metric_name = spec.metric_name
+                metric = getattr(torchmetrics, metric_name)
```
Comment (fcogidi) [inline mmlearn/tasks/classification.py:139]: Instead of this, we can create configs for torchmetric objects (using hydra_zen) and instantiate them from the config.

File: mmlearn/tasks/classification.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+            output_embeddings = outputs[query_modality.embedding]  # Predictions
+            label_index = batch[query_modality.target]  # True labels
+            names = batch[
+                NAME_KEY
+            ]  # List of dataset names for each element in the batch
+
+            # Filter indices where dataset name is part of the metric_name
+            matching_indices = [
+                i for i, name in enumerate(names) if name in dataset_name
+            ]
```
Comment (fcogidi) [inline mmlearn/tasks/classification.py:202]: There is `example_index`, dataset_index` and `example_ids` (which is a tuple of example_index and dataset_index) that can help you identify samples from the same dataset. Is the names key really necessary?

**Review by fcogidi — CHANGES_REQUESTED** [review]

**Review by afkanpour — CHANGES_REQUESTED** [review]
> Reviewed 9 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/VectorInstitute/mmlearn/17)* status: 9 of 12 files reviewed, 16 unresolved discussions (waiting on @fcogidi and @Negiiiin)

___
*[`mmlearn/datasets/core/dataset_info.py` line 1 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/17#-O7UFHv8EIP4FEJ8Adf7:-O7UFHv8EIP4FEJ8Adf8:b2fsfjl) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/2d3e2f4d44b73b4985c94a1425d1133878abd11a/mmlearn/datasets/core/dataset_info.py#L1)):*
> ```Python
> """Saving the necessary imformstion about a dataset for evaluation tasks."""
> ```

typo: information.
___
*[`mmlearn/datasets/core/dataset_info.py` line 55 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/17#-O7UGSAB3wgJ7cf2LtqZ:-O7UGSAB3wgJ7cf2Ltq_:bqbjri4) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/2d3e2f4d44b73b4985c94a1425d1133878abd11a/mmlearn/datasets/core/dataset_info.py#L55)):*
> ```Python
>         return self._class_count
> 
>     def set_class_count(self, count: int) -> None:
> ```

Do we really need this setter? I imagine that for a dataset info object we need to set this once at initialization, and it won't change after that. Are there scenarios in which this is still needed?
___
*[`mmlearn/datasets/core/dataset_info.py` line 80 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/17#-O7UI3ICEdq9y1mQM8lv:-O7UI3ICEdq9y1mQM8lw:b1ak36s) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/2d3e2f4d44b73b4985c94a1425d1133878abd11a/mmlearn/datasets/core/dataset_info.py#L80)):*
> ```Python
>         return self._label_mapping
> 
>     def set_label_mapping(self, mapping: Dict[str, Any]) -> None:
> ```

Same comment.
___
*[`mmlearn/datasets/core/dataset_info.py` line 107 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/17#-O7UIDoc-hKB1uEvewiZ:-O7UIDoc-hKB1uEvewi_:b-bx4fjw) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/2d3e2f4d44b73b4985c94a1425d1133878abd11a/mmlearn/datasets/core/dataset_info.py#L107)):*
> ```Python
>         return self._label_embedding
> 
>     def set_label_embedding(self, embedding: Dict[str, Any]) -> None:
> ```

same comment.
___
*[`mmlearn/tasks/classification.py` line 52 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/17#gh-1765233403:-O7UJdab7g9stY2woyqx:b-5mqf0r) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/2d3e2f4d44b73b4985c94a1425d1133878abd11a/mmlearn/tasks/classification.py#L52)):*
<details><summary><i>Previously, fcogidi (Franklin) wrote…</i></summary><blockquote>

I think linear probing should be a separate task. It is not really a zero-shot task (the linear classifier needs to be trained), so it shouldn't inherit from `EvaluationHooks`.
</blockquote></details>

I agree. Zero-shot classification is retrieval, whereas linear probing and regular fine-tuning classifications add a classification head on top of the encoder output. In linear-probing the encoder weights are frozen, and in fine-tuning classification all weights are fine-tuned. We should have a class for zero-shot and another for linear probing & fine-tuning.
___
*[`mmlearn/tasks/classification.py` line 70 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/17#-O7UKsIz6cDLMfsncDEA:-O7UKsJ-0jHRrYog4KjI:b-21ut08) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/2d3e2f4d44b73b4985c94a1425d1133878abd11a/mmlearn/tasks/classification.py#L70)):*
> ```Python
>         if self.mode == "zeroshot":
> 
>             class LabelDescriptionDataset(Dataset):
> ```

I'm not sure if defining classes inside conditional statements is a good software engineering practice. `on_evaluation_epoch_start` is a recurring function, keep in mind that this if statement runs multiple times. It doesn't make sense that the class definition is embedded here.
___
*[`mmlearn/tasks/classification.py` line 84 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/17#-O7ULGAFAGyD2gBkuM1S:-O7ULGAFAGyD2gBkuM1T:bh76b73) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/2d3e2f4d44b73b4985c94a1425d1133878abd11a/mmlearn/tasks/classification.py#L84)):*
> ```Python
>                     tokens = self.tokenizer(description)
> 
>                     example = Example(
> ```

Does this mean that this class only works for image-text data? Can we make the implementation more general so it works for any two modalities?
___
*[`mmlearn/tasks/classification.py` line 86 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/17#-O7UL7j26D_uoArMmv6P:-O7UL7j26D_uoArMmv6Q:b-8btxan) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/2d3e2f4d44b73b4985c94a1425d1133878abd11a/mmlearn/tasks/classification.py#L86)):*
> ```Python
>                     example = Example(
>                         {
>                             Modalities.RGB: torch.rand(3, 224, 224),
> ```

Hard-coding values like this is a big NO! Please use configs for such values.



_Code quote:_
```Python
3, 224, 224
```
___
*[`mmlearn/tasks/classification.py` line 110 at r1](https://reviewable.io/reviews/VectorInstitute/mmlearn/17#-O7ULbz_3kghxsxG7lu-:-O7ULbz_3kghxsxG7lu0:ba5w8e) ([raw file](https://github.com/VectorInstitute/mmlearn/blob/2d3e2f4d44b73b4985c94a1425d1133878abd11a/mmlearn/tasks/classification.py#L110)):*
> ```Python
>             for name, dataset_info in all_dataset_info.items():
>                 descriptions = [
>                     "This image has a sign of " + label
> ```

Hardcoding a prompt like this seems like a very specific type of zero-shot classification. Can we implement a more general class, and  perhaps introduce a notion of "prompt" which could take a value like this string?

File: mmlearn/tasks/classification.py (hunk 1 above)
Comment (Negiiiin) [reply mmlearn/tasks/classification.py]: Doen

File: mmlearn/tasks/classification.py (hunk 2 above)
Comment (Negiiiin) [reply mmlearn/tasks/classification.py]: Done

File: mmlearn/tasks/classification.py (hunk 3 above)
Comment (Negiiiin) [reply mmlearn/tasks/classification.py]: Done

File: mmlearn/tasks/classification.py (hunk 4 above)
Comment (Negiiiin) [reply mmlearn/tasks/classification.py]: Done

File: mmlearn/tasks/classification.py (hunk 5 above)
Comment (Negiiiin) [reply mmlearn/tasks/classification.py]: Done

File: mmlearn/tasks/classification.py (hunk 4 above)
Comment (Negiiiin) [reply mmlearn/tasks/classification.py]: The only thing is that to give the descriptions to the model, I also need to pass `Modalities.RGB` to the model as well (if I have not missed anything)

How about if I define a dummy image in constants?

File: mmlearn/tasks/zero_shot_classification.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        self.all_dataset_info = all_dataset_info
+
+        for dataset_name in list(all_dataset_info.keys()):
+            for spec in self.task_specs:
+                assert Modalities.has_modality(spec.query_modality)
+                query_modality = Modalities.get_modality(spec.query_modality)
+                metric_name = spec.metric_name
+                metric = getattr(torchmetrics, metric_name)
+                num_classes = all_dataset_info[dataset_name].get_class_count()
+                task = "multiclass" if num_classes > 2 else "binary"
```
Comment (fcogidi) [inline mmlearn/tasks/zero_shot_classification.py:105]: We can keep this to multiclass (hard-coded metrics) for now.

File: projects/med_benchmarking/configs/experiment/linear_probing.yaml
Comment (fcogidi) [inline projects/med_benchmarking/configs/experiment/linear_probing.yaml:1]: Please create a new PR for the linear classifier task.

File: mmlearn/tasks/zero_shot_classification.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+    Parameters
+    ----------
+    task_specs : List[ClassificationTaskSpec]
+        A list of classification task specifications.
+    """
+
+    def __init__(
+        self,
+        task_specs: List[ClassificationTaskSpec],
```
Comment (fcogidi) [inline mmlearn/tasks/zero_shot_classification.py:45]: I think this would be easier to add to the config if it were `Dict[str, ClassificationTaskSpec]`. You will likely not need `configs/experiment/zeroshot.yaml` because you can easily specify it in the command line. It's harder to add new list configs from the command line.

File: mmlearn/datasets/core/dataset_info.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
+"""Saving the necessary information about a dataset for evaluation tasks."""
+
+from typing import Any, Dict, Optional
+
+
+class DatasetInfo:
```
Comment (fcogidi) [inline mmlearn/datasets/core/dataset_info.py:6]: Instead of this class, how about defining `num_classes` and `label2id` in each dataset class (probably as a class variable), then have `CombinedDataset` merge it if it's available?

**Review by fcogidi — CHANGES_REQUESTED** [review]

File: mmlearn/tasks/linear_probing_classification.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+class ClassificationTaskSpec:
+    """Specification for a classification task."""
+
+    metric_name: str
+    query_modality: str
+    top_k: List[int]
+
+
+@store(group="eval_task", provider="mmlearn")
+class LinearProbingClassification(EvaluationHooks):
```
[the remark this answers is not in the document]
Comment (Negiiiin) [reply mmlearn/tasks/linear_probing_classification.py:28]: True, I defined a new class for it. I am not sure if I understand your comment on this one correctly.

The way I added it is that I added new heads (same size as the classes) and just consider those as logits.

File: mmlearn/tasks/zero_shot_classification.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+                                                "num_classes": num_classes,
+                                            }
+                                        )
+                                    ).to(pl_module.device)
+                                )
+                                for k in spec.top_k
+                            }
+                        )
+                    }
+                )
```
[the remark this answers is not in the document]
Comment (Negiiiin) [reply mmlearn/tasks/zero_shot_classification.py:135]: Each metric needs its own num_classes (based on each dataset) while we are defining the metric. So I don't think that we can define one metric collection and clone that for each dataset.

But I can move this whole step to __init__

is that good enough?

File: mmlearn/tasks/zero_shot_classification.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+
+            # Because model needs to also get Modalities.RGB
+            batch[Modalities.RGB] = torch.rand(len(descriptions), 3, 224, 224).to(
+                pl_module.device
+            )
+
+            with torch.no_grad():
+                processed_example = pl_module(batch)
+
+            all_dataset_info[name].set_label_embedding(
```
[the remark this answers is not in the document]
Comment (Negiiiin) [reply mmlearn/tasks/zero_shot_classification.py:92]: Each dataset has its own label_embeddings that's why it is being stored in DatasetInfo

File: mmlearn/datasets/core/dataset_info.py (hunk 10 above)
Comment (Negiiiin) [reply mmlearn/datasets/core/dataset_info.py:6]: I think this will put more work for user in defining each dataset. Right now the user only needs to define one dictionary that is for label_mapping and everything else such as number of classes and the label embeddings of that dataset will be handled with DatasetInfo
